### ISSUE TITLE
GHA tests: Fix failure output

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5236]
This is the output I got in one of my recent PRs:

```
Detailed test results
---------------------
Fail      Other      user.test         2163 UserRoleAdminTestCase->testRoleAdmi

Fail      Other      user.test         2174 UserRoleAdminTestCase->testRoleAdmi

Fail      Browser    user.test         2180 UserRoleAdminTestCase->testRoleAdmi
---- User: User role administration (UserRoleAdminTestCase) ----
Fail      Other      user.test         2182 UserRoleAdminTestCase->testRoleAdmi

Fail      Other      user.test         2183 UserRoleAdminTestCase->testRoleAdmi

Fail      Other      user.test         2185 UserRoleAdminTestCase->testRoleAdmi
Status    Group      Filename          Line Function                            
Fail      Browser    user.test         2186 UserRoleAdminTestCase->testRoleAdmi
--------------------------------------------------------------------------------
Fail      Other      user.test         2187 UserRoleAdminTestCase->testRoleAdmi
    The title on the form is "Add role".
Fail      Other      user.test         2215 UserRoleAdminTestCase->testRoleAdmi
    Redirected to correct URL after clicking the "Save role" button.
Fail      Other      user.test         2216 UserRoleAdminTestCase->testRoleAdmi
    Link Configure role does not exist on
    http://localhost/admin/config/people/permissions
    Redirected to correct URL after clicking the "Configure role" operation.
    The title on the form is "Configure role".
    "Save role" button found on "Configure role" form.
    Link Cancel does not exist on
    http://localhost/admin/config/people/permissions
    Redirected to the main roles listing page.
    The role has had its label changed.
    The role has the new label.
```

I'd expect this instead:
```
Detailed test results
---------------------
---- User: User role administration (UserRoleAdminTestCase) ----
Status    Group      Filename          Line Function                  Message
Fail      Other      user.test         2163 testRoleAdministration    The title on the form is "Add role".
Fail      Other      user.test         2174 testRoleAdministration    Redirected to correct URL after clicking the "Save role" button.
Fail      Browser    user.test         2180 testRoleAdministration    Link Configure role does not exist on http://localhost/admin/config/people/permissions
Fail      Other      user.test         2182 testRoleAdministration    Redirected to correct URL after clicking the "Configure role" operation.
Fail      Other      user.test         2183 testRoleAdministration    The title on the form is "Configure role".
Fail      Other      user.test         2185 testRoleAdministration    "Save role" button found on "Configure role" form.
Fail      Browser    user.test         2186 testRoleAdministration    Link Cancel does not exist on http://localhost/admin/config/people/permissions
Fail      Other      user.test         2187 testRoleAdministration    Redirected to the main roles listing page.
Fail      Other      user.test         2215 testRoleAdministration    The role has had its label changed.
Fail      Other      user.test         2216 testRoleAdministration    The role has the new label.
```

...or alternatively this:
```
Detailed test results
---------------------
---- User: User role administration (UserRoleAdminTestCase) ----
Status    Group      Filename          Line Function                            
Fail      Other      user.test         2163 testRoleAdministration
  The title on the form is "Add role".
Fail      Other      user.test         2174 testRoleAdministration
  Redirected to correct URL after clicking the "Save role" button.
Fail      Browser    user.test         2180 testRoleAdministration
  Link Configure role does not exist on http://localhost/admin/config/people/permissions
Fail      Other      user.test         2182 testRoleAdministration
  Redirected to correct URL after clicking the "Configure role" operation.
Fail      Other      user.test         2183 testRoleAdministration
  The title on the form is "Configure role".
Fail      Other      user.test         2185 testRoleAdministration
  "Save role" button found on "Configure role" form.
Fail      Browser    user.test         2186 testRoleAdministration
  Link Cancel does not exist on http://localhost/admin/config/people/permissions
Fail      Other      user.test         2187 testRoleAdministration
  Redirected to the main roles listing page.
Fail      Other      user.test         2215 testRoleAdministration
  The role has had its label changed.
Fail      Other      user.test         2216 testRoleAdministration
  The role has the new label.
```

**Comment by indigoxela** [comment]
Hey, in how far is your expectation different to https://github.com/backdrop/backdrop-issues/issues/5206
:smile: 

Agreed, cutting off at 80 chars makes it harder to identify the problem.

**Comment by klonos** [comment]
It's not only the 80 character limit (which is what #5206 is about). Its also about:

- the `---- User: User role administration (UserRoleAdminTestCase) ----` line not coming first.
- the repetition of the class name (the `UserRoleAdminTestCase->` prefix being added to the actual function name), which I don't find useful if we were grouping the failures by class. In other words, I believe that using these lines as separators would suffice:
  `---- User: User role administration (UserRoleAdminTestCase) ----`
- how the failure messages are not lined up with the line numbers

If all of the above are side effects of the 80 char limit, and will be fixed by removing that limit, than we can close as a duplicate.

**Comment by indigoxela** [comment]
> If all of the above are side effects of the 80 char limit, and will be fixed by removing that limit, than we can close as a duplicate.

I'm not sure, if everything you request is feasible.

On the other hand... we could close my issue and figure out what's feasible in this one. At least, now there's two people interested in an improvement of test result output.

But be warned, it's not on top of my todo list. :wink:

**Comment by klonos** [comment]
> But be warned, it's not on top of my todo list. 😉

No worries. Not my top priority either - just creating issues as a to-do list 🙂

**Comment by quicksketch** [comment]
I filed a stab at solving this problem at https://github.com/backdrop/backdrop/pull/3764

The center of the problem seems to be in `simpletest_script_print()` where we have a switch like this:

```
  // For fails and exceptions, print to the error log.
  if ($status === 'fail' || $status === 'exception') {
    fwrite(STDERR, $message);
  }
  else {
    echo $message;
  }
```

My theory is that GitHub Actions prints error output after all the normal standard output. So one solution would be to pipe `stderr` to `stdout` by appending `&2>1` to the `run-tests.sh` command.

**Comment by quicksketch** [comment]
I pushed an additional update which increases the character wrap limit from 80 characters to 120 characters.

**Comment by quicksketch** [comment]
The PR solves most of the requested functionality. Here's a current screenshot:
![image](https://user-images.githubusercontent.com/100206/135567767-9f403ae0-60e9-4665-9fc0-3e2c85555874.png)

I'm not sure about removing the class name from each line. Sure it's redundant, but if we increase the line limit to 120 characters, there's no need to be stingy about the column width.

For the message output: I think we should continue to print that on a separate line rather than trying to inline everything. Some messages already exceed 120 character just by themselves, in which case the message is wrapped at the line limit.

**Comment by indigoxela** [comment]
@quicksketch many thanks for digging into this, this is a *huge* improvement!

One question: when there are several test failures - are they all grouped by test?

**Comment by quicksketch** [comment]
> One question: when there are several test failures - are they all grouped by test?

I don't quite know what you mean here. If what you're asking is: "if a test class contains multiple test methods (prefixed with `test*`) are they grouped by class?" If that's the question, then yes, each class creates it's own group. So repeating the class name is entirely redundant.

Annotating my previous screenshot:

![image](https://user-images.githubusercontent.com/100206/135584469-75ef0097-3414-43a2-9af9-caf136ff6597.png)

But within each group, it's definitely possible that the function/method name will be different per failure.

**Comment by indigoxela** [comment]
>> One question: when there are several test failures - are they all grouped by test?

> I don't quite know what you mean here.

Let me rephrase it... :wink: 

If different *test classes* have failures, are they grouped by class ("heading", then related failure listing, next "heading" related failure listing...)?

Your example fails in UserRoleAdminTestCase. What would it look like if, for instance, CommentNodeChangesTestCase also had failures?

**Comment by quicksketch** [comment]
> Your example fails in UserRoleAdminTestCase. What would it look like if, for instance, CommentNodeChangesTestCase also had failures?

Then it would be in under a separate "table" with a heading dividing the table from the previous test. So yes each test case is separated. There can be multiple test functions/methods within a single class though.

**Comment by quicksketch** [comment]
I've updated the PR, now with no redundant class name within each row of result output.

**Comment by klonos** [comment]
I checked out the (intentional) PR failures and I like the output 👍🏼

**Comment by quicksketch** [comment]
Here is a screenshot of the current output:

![image](https://user-images.githubusercontent.com/100206/136321975-e77cf830-5fda-44c5-9e45-18e0cf9e2383.png)

I've updated the PR to remove the intentional failures, so this is now ready for final review.

**Comment by indigoxela** [comment]
@quicksketch many thanks for fixing the failure output and for cleaning up the script code. RTBC :+1:

**Comment by quicksketch** [comment]
Merged https://github.com/backdrop/backdrop/pull/3764 into 1.x and 1.20.x. Thanks @indigoxela and @klonos!
